common/remote/sockfile.cpp

```diff
@@ -3811,17 +3832,247 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
+    unsigned __int64 chooseN = 0;
+    IArrayOf<IKeySegmentMonitor> segMonitors;
+    Owned<ISourceRowPrefetcher> prefetcher;
+    CThorContiguousRowBuffer prefetchBuffer;
+    bool opened = false;
+    bool eofseen = false;
+    bool needTransform = true;
+    unsigned __int64 processed = 0;
+    Owned<ISerialStream> inputstream;
+    Owned<IFileIO> iFileIO;
+    unsigned __int64 startPos = 0;
+    Linked<IRtlFieldTypeDeserializer> inputFieldInfoDeserializer;
+    Linked<IRtlFieldTypeDeserializer> outputFieldInfoDeserializer;
+    Linked<IOutputMetaData> outMeta;
+
+    void open()
+    {
+        if (!helper->canMatchAny())
+            eofseen = true;
+        else
+        {
+            const char *fileName = helper->getFileName();
+
+            OwnedIFile iFile = createIFile(fileName);
+
+#if 0
+            bool compressed = false; // isCompressedFile(iFileIO); // Should be passed with JSON
+            StringBuffer encryptionkey;
+            if (compressed)
+            {
+                Owned<IExpander> eexp;
+                if (encryptionkey.length()!=0)
+                    eexp.setown(createAESExpander256((size32_t)encryptionkey.length(),encryptionkey.bufferBase()));
+                iFileIO.setown(createCompressedFileReader(iFile,eexp));
+                if(!iFileIO && !blockcompressed) //fall back to old decompression, unless dfs marked as new
+                {
+                    iFileIO.setown(iFile->open(IFOread));
+                    if(iFileIO)
+                        rowcompressed = true;
+                }
+            }
+            else
+#endif
+                iFileIO.setown(iFile->open(IFOread));
+            if (!iFileIO)
+                throw MakeStringException(0, "Failed to open: '%s'", fileName);
+
+            inputstream.setown(createFileSerialStream(iFileIO, startPos));
+            prefetchBuffer.setStream(inputstream);
+            prefetcher.setown(helper->queryDiskRecordSize()->createDiskPrefetcher(nullptr, 0));
+
+            outBuilder = new MemoryBufferBuilder(resultBuffer, helper->queryOutputMeta()->getMinRecordSize());
+            chooseN = helper->getChooseNLimit();
+            limit = helper->getRowLimit();
+        }
+
+        opened = true;
+    }
+    void close()
+    {
+        // TBD
+    }
+    bool segMonitorsMatch(const void *row) { return true; }
+public:
+    CRemoteDiskReadActivity(IHThorDiskReadArg &_helper, IRtlFieldTypeDeserializer &_inputFieldInfoDeserializer, IRtlFieldTypeDeserializer &_outputFieldInfoDeserializer)
+        : helper(&_helper), inputFieldInfoDeserializer(&_inputFieldInfoDeserializer), outputFieldInfoDeserializer(&_outputFieldInfoDeserializer), prefetchBuffer(nullptr)
+    {
+        /* NB: this objects links the IRtlFieldTypeDeserializer's, purely because RtlRecordTypeInfo that have been deserialized are used
+         * by this class and the helper, and would become invalid if the deserializer objects were destroyed.
+         */
```

Comment:
@ghalliday - is there a cleaner alternative ? Seems a shame to have these members, solely for this reason.

As I mentioned before, to me it looks like getTypeInfo() that get a RtlRecordTypeInfo info from the IRtlFieldTypeDeserializer would be safe to use beyond releasing the IRtlFieldTypeDeserializer, e.g. if RtlRecordTypeInfo objects linked the IRtlFieldTypeDeserializer, then this wouldn't be necessary.

Reply:
The disk read activity shouldn't know anything about the field information deserializer.  That should be manged by the helper.  Possibly needs a derived version, or a query object.

Reply:
I've moved the deserializers into a variety of the helper.